src/models/RecordHoldingsMap.ts

```diff
@@ -0,0 +1,85 @@
+import { Result } from '../interfaces/Result';
+import { BibdataService } from '../services/BibdataService';
+import { PhysicalHolding } from './PhysicalHolding';
+import { SearchResults } from './SearchResults';
+
+type RecordAndHoldings = { [recordId: string]: PhysicalHolding[] };
+
+export class RecordHoldingsMap {
+  mapping: RecordAndHoldings;
+  constructor(results: SearchResults) {
+    this.mapping = Object.fromEntries(
+      results.records.map(document => {
+        return [document.id, this.#extractHoldings(document)];
+      })
+    );
+  }
+
+  getHoldingsByDocumentId(id: string): PhysicalHolding[] {
+    return this.mapping[id];
+  }
+
+  updateScsbAvailability(): Promise<boolean> {
+    const bibdata = new BibdataService();
+    return bibdata
+      .scsbAvailability(this.barcodes())
+      .then(results => {
+        Object.keys(results).forEach(barcode => {
+          Object.keys(this.mapping).forEach(recordId => {
+            this.mapping[recordId].forEach(holding => {
+              if (holding.barcode === barcode) {
+                holding.status = results[barcode].status;
+              }
+            });
+          });
+        });
+        return true;
+      })
+      .catch(() => {
+        Object.keys(this.mapping).forEach(recordId => {
+          if (recordId.startsWith('SCSB-')) {
+            this.mapping[recordId].forEach(holding => {
+              holding.status = 'Check record for full availability';
+            });
+          }
+        });
+        return true;
+      });
+  }
+
+  barcodes(): string[] {
```

Comment:
I wanted to make this method private, but it failed in the browser when I tried it. 😒